public/app/features/dashboard-scene/utils/clone.ts

```diff
@@ -0,0 +1,73 @@
+const CLONE_KEY = '-clone-';
+const CLONE_SEPARATOR = '|';
+
+const CLONED_KEY_REGEX = new RegExp(`${CLONE_KEY}[1-9]+$`);
+const ORIGINAL_REGEX = new RegExp(`${CLONE_KEY}\\d+$`);
+
+/**
+ * Create or alter the last key for a key
+ * @param key
+ * @param index
+ */
+export function getCloneKey(key: string, index: number): string {
+  const parts = key.split(CLONE_SEPARATOR).slice(0, -1);
+  const lastKey = getOriginalKey(getLastKeyFromClone(key));
+  return [...parts, `${lastKey}${CLONE_KEY}${index}`].join(CLONE_SEPARATOR);
+}
+
+/**
+ * Get the original key from a clone key
+ * @param key
+ */
+export function getOriginalKey(key: string): string {
+  return getLastKeyFromClone(key).replace(ORIGINAL_REGEX, '');
+}
+
+/**
+ * Checks if the last key is a clone key
+ * @param key
+ */
+export function isClonedKey(key: string): boolean {
+  return CLONED_KEY_REGEX.test(getLastKeyFromClone(key));
+}
+
+/**
+ * Checks if key1 is a clone of key2
+ * @param key1
+ * @param key2
+ */
+export function isClonedKeyOf(key1: string, key2: string): boolean {
+  return isClonedKey(key1) && getOriginalKey(key1) === getOriginalKey(key2);
+}
+
+/**
+ * Checks if the key or any of its ancestors are cloned
+ * @param key
+ */
+export function isInCloneChain(key: string): boolean {
+  return key.split(CLONE_SEPARATOR).some(isClonedKey);
+}
+
+/**
+ * Get the last key from a clone key
+ * @param key
+ */
+export function getLastKeyFromClone(key: string): string {
+  return key.split(CLONE_SEPARATOR).pop() ?? '';
+}
+
+/**
+ * Join clone keys
+ * @param keys
+ */
+export function joinCloneKeys(...keys: string[]): string {
+  return keys.filter(Boolean).join(CLONE_SEPARATOR);
```

Comment:
This is cool!

Reply:
what does it do?

Reply:
Given a set of keys, this helper method allows you to join them in a hierarchy instead of manually doing: `${key1}${CLONE_SEPARATOR}${key2}${CLONE_SEPARATOR}${key3}` etc. We also filter out the empty strings by using `filter(Boolean)`.

Reply:
I remarked the `.filter(Boolean)` usage here, never seen that before and learned something new after reviewing.